streams/shared/src/main/scala/zio/stream/ZStream.scala

```diff
@@ -1792,6 +1798,35 @@ class ZStream[-R, +E, +A](val process: ZManaged[R, E, Pull[R, E, A]]) extends Se
    */
   final def zipWithIndex: ZStream[R, E, (A, Int)] =
     mapAccum(0)((index, a) => (index + 1, (a, index)))
+
+  /**
+   * Zips the two streams so that when a value is emitted by either of the two streams,
+   * it is combined with the latest value from the other stream to produce a result.
+   */
+  final def zipWithLatest[R1 <: R, E1 >: E, B, C](that: ZStream[R1, E1, B])(f0: (A, B) => C): ZStream[R1, E1, C] =
+    ZStream[R1, E1, C] {
+      for {
+        is    <- self.mergeEither(that).process
+        state <- Ref.make[(Option[A], Option[B])]((None, None)).toManaged_
+        pull: Pull[R1, E1, C] = {
```

Comment:
Nit: you could just put the `pull` expression directly in the `yield`.

Reply:
It's kind of a pattern in the `ZStream` class, I think it's a little easier to read.